Fix lag issue with Google Maps when zooming

agpixdev opened the pull request:
The fix is to force the view center right after zooming. What I have observed is that under some circumstances, Google Maps will delay the zoom operation until it sees a new center being set or a timeout elapses. Forcing the view center avoids the lag observed before the zoom is applied (i.e. avoids the timeout path).

agpixdev committed:
Fix lag issue with Google Maps when zooming

The fix is to force the view center right after zooming. What I have observed is that under some circumstances, Google Maps will delay the zoom operation until it sees a new center being set or a timeout elapses. Forcing the view center avoids the lag observed before the zoom is applied (i.e. avoids the timeout path).

adube reviewed (changes requested):
Code is clean.  However, I notice that the map kinda "shifts" after right before the zooming in/out occurs.

To the eye, I think there's less lag indeed, but the fact that the map kinda "shifts" makes it not that pretty.

Are you able to see what I'm talking about?

Thanks for the patch!

agpixdev commented:
I'll take a look at it tomorrow to see what is the effect you are describing. Thanks !

agpixdev commented:
After looking at the "shift" effect in the example, I went back to the original development commits and realized that the issue had been fixed in two steps:

1. Preparing the setZoomAndCenter() fix in ol3-google-maps (as submitted)
2. Setting the OpenLayers MouseWheelZoom interaction configuration options to have zoomDuration = 0 in the application

It looks like the second change actually gets rid of both effects; the initial lag and the "shift" effect. So this is an OpenLayers-induced behavior that can be fixed by configuration.

Conclusion: please disregard this pull-request completely.

adube commented:
Thanks for the explanation.

agpixdev committed:
Merge branch 'master' into zoom-fix-googlemaps-lag-pr

agpixdev committed:
Remove 'shift' effect when zooming with mouse wheel

Set the mouse wheel zoom animation duration to 0 to remove the 'shift' effect visible while zooming.

agpixdev commented:
I'm reopening this pull-request for the following reason:

The original lag this is fixing is still present event if we set the mouse wheel zoom animation duration to 0.
The lag this is fixing occurs in the following way:

1. First zoom the map (mouse wheel zooming used)
2. Drag the map. You will notice a lag before the map starts to move. This only happens for the first drag after a zoom.
3. The fix proposed here removes this lag on the first drag following a zoom.

The lag can be observed in the concept example by repeating steps 1 & 2 above. With the fix proposed, the lag disappears.

Sorry if my previous explanation was a little bit off (this change was done many months ago, had to retest it to see the real cause for the fix).

agpixdev commented:
New description should be:

The fix is to force the view center right after zooming. What I have observed is that under some circumstances, Google Maps will delay the first drag operation after a zoom until it sees a new center being set or a timeout elapses. Forcing the view center avoids the lag observed before the drag is applied (i.e. avoids the timeout path).

adube commented:
Alright, I'll test the patch this weekend.

JamiesWhiteShirt commented on src/olgm/interaction.js:
```diff
@@ -33,6 +33,7 @@ export function defaults(opt_options) {
   options['altShiftDragRotate'] = false;
   options['dragPan'] = false;
   options['pinchRotate'] = false;
+  options['zoomDuration'] = 0;
```
It would be nice if this was documented in the jsdoc above.

agpixdev replied:
Agree. Updated !

agpixdev committed:
Updated jsdoc for proposed change

adube commented:
Code is clean.  The animation runs smoothly now.  Thank you very much for this patch!
